iib/web/models.py

```diff
@@ -190,6 +204,79 @@ def add_state(self, state, state_reason):
         db.session.flush()
         self.request_state_id = request_state.id
 
+    def to_json(self, verbose=True):
+        states = [
+            {
+                'state': RequestStateMapping(state.state).name,
+                'state_reason': state.state_reason,
+                'updated': state.updated.isoformat(),
+            }
+            for state in self.states
+        ]
+        # Reverse the list since the latest states should be first
+        states = list(reversed(states))
+        latest_state = states[0]
+
+        rv = {
+            'id': self.id,
+            'bundles': [bundle.to_json() for bundle in self.bundles],
+            'binary_image': self.binary_image.to_json(),
+            'from_index': self.from_index.to_json(),
+            'user': self.user.username,
+            'latest_state': latest_state
+        }
+
+        if verbose:
+            rv['state_history'] = states
+
+        return rv
+
+    @classmethod
+    def from_json(cls, kwargs):
+        # Validate all required parameters are present
+        required_params = {'bundles', 'binary_image'}
+        optional_params = {'from_index', 'add_arches', 'cnr_token'}
+        missing_params = required_params - set(kwargs.keys()) - optional_params
+        if missing_params:
+            raise ValidationError('Missing required parameter(s): {}'
+                                  .format(', '.join(missing_params)))
+
+        # Don't allow the user to set arbitrary columns or relationships
+        invalid_params = set(kwargs.keys()) - required_params - optional_params
+        if invalid_params:
+            raise ValidationError(
+                'The following parameters are invalid: {}'.format(', '.join(invalid_params)))
+
+        request_kwargs = deepcopy(kwargs)
+
+        # Validate bundles are correctly provided
+        bundles = request_kwargs.get('bundles')
+        if not isinstance(bundles, list) or len(bundles) == 0:
+            raise ValidationError('"bundles" should be a non-empty array')
+
+        # Validate binary_image is correctly provided
+        if not isinstance(request_kwargs.get('binary_image'), str):
```

Comment:
`request_kwargs['binary_image']` would need to be an `Image` object before it is passed in `request = cls(**request_kwargs)`